runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/state/FlinkStateInternals.java

```diff
@@ -139,17 +163,27 @@ private FlinkStateBinder(StateNamespace namespace, StateContext<?> stateContext)
     @Override
     public <T2> ValueState<T2> bindValue(
         String id, StateSpec<ValueState<T2>> spec, Coder<T2> coder) {
-      return new FlinkValueState<>(flinkStateBackend, id, namespace, coder);
+      ValueStateDescriptor<T2> valueStateDescriptor =
+          new ValueStateDescriptor<>(id, new CoderTypeSerializer<>(coder));
+      globalWindowStateDescriptors.add(valueStateDescriptor);
+      return new FlinkValueState<>(flinkStateBackend, id, namespace, valueStateDescriptor);
     }
 
     @Override
     public <T2> BagState<T2> bindBag(String id, StateSpec<BagState<T2>> spec, Coder<T2> elemCoder) {
-      return new FlinkBagState<>(flinkStateBackend, id, namespace, elemCoder);
+      ListStateDescriptor<T2> listStateDescriptor =
+          new ListStateDescriptor<>(id, new CoderTypeSerializer<>(elemCoder));
+      globalWindowStateDescriptors.add(listStateDescriptor);
```

Comment:
This is executed on every state access in the portable runner. It might be possible to reuse the descriptors. How much could be the net benefit? Probably small compared to the fn api overhead.

Reply:
This is unrelated to the changes here. This object was just created a layer down before (FlinkBagState). Let's handle such optimizations in a follow-up.

Reply:
Changed reverted.